### PR TITLE
Property to terminate tbb threads

### PR DESCRIPTION
**Pull request by riverlijunjie**
### Details:
 - Property to force terminate tbb threads
 - Change frontendManager and executorManager to be non static instance

### Tickets:
 - 82487

### CODE REVIEW

**Review by wangleis — APPROVED** [review]

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp [hunk 1]
```diff
@@ -385,6 +386,18 @@ TEST_P(OVClassSetDevicePriorityConfigTest, SetConfigAndCheckGetConfigNoThrow) {
     ASSERT_EQ(devicePriority, configuration[ov::device::priorities.name()].as<std::string>());
 }
 
+TEST_P(OVClassSetTBBForceTerminatePropertyTest, SetConfigNoThrow) {
+    ov::Core ie = createCoreWithTemplate();
+
+    bool value = true;
+    OV_ASSERT_NO_THROW(ie.set_property(deviceName, ov::force_tbb_terminate(false)));
```
Comment (wangleis) [inline src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:393]: force_tbb_terminate is property of OV runtime. Please remove device name in test case.

File: src/inference/src/ie_core.cpp [hunk 2]
```diff
@@ -1113,6 +1123,18 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
     void SetConfigForPlugins(const std::map<std::string, std::string>& configMap, const std::string& deviceName) {
         auto config = configMap;
 
+        for (auto& item : config) {
+            if (item.first == ov::force_tbb_terminate.name()) {
```
Comment (wangleis) [inline src/inference/src/ie_core.cpp:1127]: Why is force_tbb_terminate one property of pluign instead of OV runtime?

Comment (riverlijunjie) [reply src/inference/src/ie_core.cpp:1127]: This property should take effect for all plugins that involved tbb threads.

File: src/frontends/common/src/manager.cpp [hunk 3]
```diff
@@ -242,3 +242,40 @@ template <>
...
     return load_by_model_impl(variants);
 }
+namespace ov {
+namespace frontend {
+
+namespace {
+
+class FrontEndManagerHolder {
+    std::mutex m_mutex;
+    std::weak_ptr<FrontEndManager> m_manager;
```
Comment (tomdol) [inline src/frontends/common/src/manager.cpp:252]: Is it a weak pointer intentionally? Should the FrontEndManager objects created in this pattern be able to be destroyed and then replaced with other instances if needed?

File: src/inference/include/openvino/runtime/properties.hpp [hunk 4]
```diff
@@ -509,6 +509,15 @@ static constexpr Property<uint32_t, PropertyMutability::RW> auto_batch_timeout{"
...
     range_for_async_infer_requests{"RANGE_FOR_ASYNC_INFER_REQUESTS"};
 
+/**
+ * @brief Read-write property to set whether force terminate tbb when ov core destruction
+ * value type: boolean
+ *   - True explicitly terminate tbb when ov core destruction
+ *   - False will not involve additional tbb operations when core destruction
+ * @ingroup ov_runtime_cpp_prop_api
+ */
+static constexpr Property<bool, PropertyMutability::RW> force_tbb_terminate{"FORCE_TBB_TERMINATE"};
```
Comment (jiwaszki) [inline src/inference/include/openvino/runtime/properties.hpp:519]: Please add bindings to it here:
https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/pyopenvino/core/properties/properties.cpp

Without it Python API will lack new property. Tests for it could be useful as well.

**Review by jiwaszki — CHANGES_REQUESTED** [review]

File: src/core/src/op/constant.cpp [hunk 5]
```diff
@@ -14,6 +14,7 @@
 #include "ngraph/log.hpp"
 #include "ngraph/op/util/attr_types.hpp"
 #include "ngraph/util.hpp"
+#include "openvino/frontend/manager.hpp"
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/src/op/constant.cpp:17]: It makes sense, actually I have adopt this solution for ov::model, but not notice here. I will follow your suggestion.

File: src/core/include/openvino/op/constant.hpp [hunk 6]
```diff
@@ -716,6 +716,7 @@ class OPENVINO_API Constant : public Op {
     element::Type m_element_type;
     Shape m_shape{};
     std::shared_ptr<ngraph::runtime::AlignedBuffer> m_data;
+    std::shared_ptr<ov::frontend::FrontEndManager> m_femgr;
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/include/openvino/op/constant.hpp:719]: Actually, I only found op::constant will trigger segment fault during its destruction, so only put the dependency into op::constant. Ok let me check whether it can work if move to node:)

File: src/inference/src/ie_core.cpp [hunk 7]
```diff
@@ -34,6 +35,7 @@
 #include "ngraph/opsets/opset.hpp"
 #include "ngraph/pass/constant_folding.hpp"
 #include "openvino/core/except.hpp"
+#include "openvino/frontend/manager.hpp"
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/inference/src/ie_core.cpp:38]: ok

File: src/frontends/common/src/manager.cpp (hunk 3 above)
Comment (riverlijunjie) [reply src/frontends/common/src/manager.cpp:252]: yes, it is designed as a weak pointer so that it can be destroyed during openvino dll/lib unloaded, ov::core also take a reference of it so that it will not be released before ov::core destruction.

File: src/inference/include/openvino/runtime/properties.hpp (hunk 4 above)
Comment (riverlijunjie) [reply src/inference/include/openvino/runtime/properties.hpp:519]: sure, will do it!

Comment (riverlijunjie) [reply src/inference/include/openvino/runtime/properties.hpp:519]: done

File: src/core/src/op/constant.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/src/op/constant.cpp]: done

File: src/core/include/openvino/op/constant.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/include/openvino/op/constant.hpp]: Have applied void for it: std::shared_ptr<void> m_femgr;

File: src/inference/src/ie_core.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/inference/src/ie_core.cpp]: done

**Review by jiwaszki — COMMENTED** [review]
Python side is OK.

**Review by t-jankowski — CHANGES_REQUESTED** [review]
`OV Core` should be independent to any `FrontEnd`.

File: src/core/include/openvino/core/model.hpp [hunk 8]
```diff
@@ -114,6 +115,8 @@ class OPENVINO_API Model : public std::enable_shared_from_this<Model> {
     /// \brief Clones the original model
     std::shared_ptr<ov::Model> clone() const;
 
+    std::shared_ptr<frontend::FrontEndManager> m_femgr;
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/include/openvino/core/model.hpp]: 1. In latest commit, it has been changed to: std::shared_ptr<void> m_femgr;
2. Some description has been put.

File: src/inference/src/threading/ie_executor_manager.cpp [hunk 9]
```diff
@@ -90,9 +149,33 @@ void ExecutorManagerImpl::clear(const std::string& id) {
...
 }
 
+namespace {
+
+class ExecutorManagerHolder {
+    std::mutex _mutex;
+    std::weak_ptr<ExecutorManager> _manager;
+
+    ExecutorManagerHolder(const ExecutorManagerHolder&) = delete;
+    ExecutorManagerHolder& operator=(const ExecutorManagerHolder&) = delete;
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/inference/src/threading/ie_executor_manager.cpp:159]: It makes sense, will update.

[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/inference/src/threading/ie_executor_manager.cpp:160]: done

File: src/core/CMakeLists.txt [hunk 10]
```diff
@@ -113,6 +114,7 @@ set_source_files_properties("${CMAKE_CURRENT_SOURCE_DIR}/src/pass/convert_precis
 
 # Defines macro in C++ to load backend plugin
 target_include_directories(ngraph_obj PUBLIC $<BUILD_INTERFACE:${OV_CORE_INCLUDE_PATH}>
+				      PUBLIC $<BUILD_INTERFACE:${OV_FRONTEND_INCLUDE_PATH}>
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/CMakeLists.txt]: done

**Review by tomdol — APPROVED** [review]
I think at this point I'm ok with the solution. Any possible future problems can be fixed iteratively.

**Review by jiwaszki — APPROVED** [review]

File: src/cmake/ie_parallel.cmake [hunk 11]
```diff
@@ -113,6 +117,7 @@ function(set_ie_threading_interface_for TARGET_NAME)
         if (TBB_FOUND)
             set(IE_THREAD_DEFINE "IE_THREAD_TBB")
             ie_target_link_libraries(${TARGET_NAME} ${LINK_TYPE} ${TBB_IMPORTED_TARGETS})
+            target_compile_definitions(${TARGET_NAME} ${COMPILE_DEF_TYPE} TBB_PREVIEW_WAITING_FOR_WORKERS=1)
```
Comment (ilya-lavrenov) [inline src/cmake/ie_parallel.cmake:120]: I suppose we should duplicate this to `ie_parallel.hpp` since if the customer does not use cmake, such definition is lost.

File: src/inference/src/ie_core.cpp [hunk 12]
```diff
@@ -258,6 +264,8 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
         }
     };
 
+    std::shared_ptr<void> frontEndManagerPtr;
```
Comment (ilya-lavrenov) [inline src/inference/src/ie_core.cpp:267]: why do we need a reference to frontend manager in core?

File: src/inference/src/threading/ie_executor_manager.cpp [hunk 13]
```diff
@@ -4,36 +4,94 @@
...
+#if IE_THREAD == IE_THREAD_TBB || IE_THREAD == IE_THREAD_TBB_AUTO
+        try {
+            if (tbbTaskScheduler && tbbThreadsCreated) {
+                tbbTaskScheduler->terminate();
+            }
+            tbbThreadsCreated = false;
+            tbbTaskScheduler = nullptr;
+        } catch (std::exception& e) {
+            tbbTaskScheduler = nullptr;
+            IE_THROW() << e.what();
```
Comment (ilya-lavrenov) [inline src/inference/src/threading/ie_executor_manager.cpp:82]: throwing exceptions from dctors does not seem a good idea.

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp (hunk 1 above)
Comment (ilya-lavrenov) [inline src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:393]: I suppose this property is NOT related to specific device and must be used as `ie.set_property( ov::force_tbb_terminate(false));`

In this case we don't need changes in `SetConfigForPlugins`

File: src/inference/src/ie_core.cpp [hunk 14]
```diff
@@ -860,6 +870,10 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
                         "You can only get_config of the AUTO itself (without devices). "
                         "get_config is also possible for the individual devices before creating the AUTO on top.");
 
+        if (name == ov::force_tbb_terminate.name()) {
+            const auto flag = executorManager()->getTbbFlag();
+            return decltype(ov::force_tbb_terminate)::value_type(flag);
+        }
```
Comment (ilya-lavrenov) [inline src/inference/src/ie_core.cpp:876]: I suppose for such needs we need to introduce `get_property` without device name, since such property is Core specific, not device-specific.
Can we do this?

File: src/core/src/model.cpp [hunk 15]
```diff
@@ -34,6 +34,13 @@ BWDCMP_RTTI_DEFINITION(ov::AttributeAdapter<std::shared_ptr<ov::Model>>);
 
 atomic<size_t> ov::Model::m_next_instance_id(0);
 
+namespace ov {
+namespace frontend {
+class FrontEndManager;
+std::shared_ptr<FrontEndManager> get_frontend_manager();
```
Comment (ilya-lavrenov) [inline src/core/src/model.cpp:40]: am I right that because of static frontend manager, we have issue with dynamically loaded OpenVINO.so:
1. FE stores reference to OpenVINO.so
2. OpenVINO stores references to frontend libraries
And because of that, we have cyclic dependencies and libraries are not unloaded?

File: src/tests/functional/inference_engine/core_threading.cpp [hunk 16]
```diff
@@ -75,6 +75,9 @@ TEST_F(CoreThreadingTests, SetConfigPluginDoesNotExist) {
     }, 10000);
 }
 
+// TODO: CVS-68982
+#ifndef OPENVINO_STATIC_LIBRARY
+
 // tested function: RegisterPlugin
 TEST_F(CoreThreadingTests, RegisterPlugin) {
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/tests/functional/inference_engine/core_threading.cpp:82]: Static test should not register plugin, we can see CVS-68982

File: src/inference/src/ie_core.cpp (hunk 12 above)
Comment (riverlijunjie) [reply src/inference/src/ie_core.cpp:267]: oh, in early solution we put a reference in ie::core to make it not released after read_model(), now maybe we can remove it from ie::core, let me verify it.

File: src/inference/src/threading/ie_executor_manager.cpp (hunk 13 above)
Comment (riverlijunjie) [reply src/inference/src/threading/ie_executor_manager.cpp:82]: Good finding! 
It has been resolved in another PR https://github.com/openvinotoolkit/openvino/pull/11964

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp (hunk 1 above)
Comment (riverlijunjie) [reply src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:393]: Yes, It has been resolved in https://github.com/openvinotoolkit/openvino/pull/12022

File: src/inference/src/threading/ie_executor_manager.cpp [hunk 17]
```diff
@@ -58,6 +116,7 @@ IStreamsExecutor::Ptr ExecutorManagerImpl::getIdleCPUStreamsExecutor(const IStre
                 return executor;
     }
     auto newExec = std::make_shared<CPUStreamsExecutor>(config);
+    tbbThreadsCreated = true;
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/inference/src/threading/ie_executor_manager.cpp:119]: We found if not thread created but call terminate api, it will cause random crash.

File: src/inference/src/ie_core.cpp (hunk 14 above)
Comment (riverlijunjie) [reply src/inference/src/ie_core.cpp:876]: Yes, It has been resolved in https://github.com/openvinotoolkit/openvino/pull/12022

File: src/core/src/model.cpp (hunk 15 above)
Comment (riverlijunjie) [reply src/core/src/model.cpp:40]: yes, it is the reason that why libopenvino_ir_frontend.so cannot be unloaded. But for libopenvino.so unload issue, there are more static variables and other dependencies/reference.

File: src/core/include/openvino/core/model.hpp [hunk 18]
```diff
@@ -114,6 +114,9 @@ class OPENVINO_API Model : public std::enable_shared_from_this<Model> {
     /// \brief Clones the original model
     std::shared_ptr<ov::Model> clone() const;
 
+    /// \brief Reference to frontend maanager
+    std::shared_ptr<void> m_femgr;
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/include/openvino/core/model.hpp:118]: It is used to keep frontend manager instance is not released before ov::Model, due to frontend manager has been work as a dynamic instance rather than static instance.

File: src/core/include/openvino/op/constant.hpp [hunk 19]
```diff
@@ -716,6 +716,7 @@ class OPENVINO_API Constant : public Op {
     element::Type m_element_type;
     Shape m_shape{};
     std::shared_ptr<ngraph::runtime::AlignedBuffer> m_data;
+    std::shared_ptr<void> m_femgr;
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/include/openvino/op/constant.hpp:719]: It is used to keep frontend manager instance is not released before ov::op::Constant which has shared memory.

File: src/frontends/common/src/manager.cpp [hunk 20]
```diff
@@ -242,3 +242,40 @@ template <>
...
+    return frontend_manager_holder;
+}
+
+}  // namespace
+
+FrontEndManager::Ptr get_frontend_manager() {
+    return get_frontend_manager_holder().get();
+}
+
+}  // namespace frontend
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/frontends/common/src/manager.cpp:280]: Not direct related with TBB terminate. We put them together because both are related to DLL unloaded issue.

File: src/inference/src/ie_core.cpp [hunk 21]
```diff
@@ -1113,6 +1127,18 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+        for (auto& item : config) {
+            if (item.first == ov::force_tbb_terminate.name()) {
+                auto flag = item.second == CONFIG_VALUE(YES) ? true : false;
+                executorManager()->setTbbFlag(flag);
+                config.erase(item.first);
+                break;
+            }
+        }
+        if (config.empty()) {
+            return;
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/inference/src/ie_core.cpp:1139]: If there is empty config, we don't need pass down anymore, it prevents passing empty config into plugins.

File: src/inference/src/ie_core.cpp (hunk 14 above)
Comment (ilya-lavrenov) [reply src/inference/src/ie_core.cpp:876]: In that PR you use `set_property` with no device name, but I think we can also add `get_property` with no device name to have mirrored API.

File: src/core/include/openvino/core/model.hpp (hunk 18 above)
Comment (ilyachur) [inline src/core/include/openvino/core/model.hpp:118]: Why do we have this definition in public place?

File: src/core/include/openvino/op/constant.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (ilyachur) [reply src/core/include/openvino/op/constant.hpp:719]: `ov::op::Constant` should be hold by `ov::Model` I don't understand why we need to duplicate frontends here

File: src/core/src/op/constant.cpp [hunk 22]
```diff
@@ -225,9 +235,14 @@ ov::op::v0::Constant::Constant(const Constant& other, const ov::Shape& new_shape
     m_data = other.m_data;
     update_identical_flags(other.m_all_elements_bitwise_identical_checked, other.m_all_elements_bitwise_identical);
     constructor_validate_and_infer_types();
+    m_femgr = ov::frontend::get_frontend_manager();
 }
 
-ov::op::v0::Constant::~Constant() = default;
+ov::op::v0::Constant::~Constant() {
```
Comment (ilyachur) [inline src/core/src/op/constant.cpp:241]: Looks like default destructor does the same things

File: src/inference/src/ie_core.cpp (hunk 12 above)
Comment (ilya-lavrenov) [reply src/inference/src/ie_core.cpp:267]: I suppose the reference here is just to prevent reloading of frontend manager each time, since if the model is loaded and then destroyed, the FEM is removed.
Keeping FEM here, will extend lifetime of FEM to lifetime of Core object. So, it's a kind of optimization  I suppose

File: src/core/src/model.cpp (hunk 15 above)
Comment (riverlijunjie) [reply src/core/src/model.cpp:40]: There are quit a lot such datas, not only static variable but also dependences data, I have ever setup a debug environment but not save the data, maybe I can re-setup and share the data to you later.

File: src/core/include/openvino/core/model.hpp (hunk 18 above)
Comment (riverlijunjie) [reply src/core/include/openvino/core/model.hpp:118]: It should be move to protected or private place.

File: src/core/include/openvino/op/constant.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/include/openvino/op/constant.hpp:719]: Maybe it is duplicated, because we first resolve the segment fault for ov::op::Constant caused by dynamic frontend manager, and then found ov::Model also exit such problem, so add reference to ov::Model. I will check/verify whether can remove reference in ov::op::Constant.

File: src/cmake/ie_parallel.cmake (hunk 11 above)
Comment (riverlijunjie) [reply src/cmake/ie_parallel.cmake:120]: ok, let me check how to solve it.

File: src/inference/src/ie_core.cpp (hunk 12 above)
Comment (riverlijunjie) [reply src/inference/src/ie_core.cpp:267]: Thanks your explanation :)

File: src/core/src/op/constant.cpp (hunk 22 above)
Comment (riverlijunjie) [reply src/core/src/op/constant.cpp:241]: But we need keep m_data be released before frontend Manager.

File: src/core/include/openvino/core/model.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (ilyachur) [reply src/core/include/openvino/core/model.hpp:118]: In general we don't need to have FR manager here. We need to keep only one frontend which created this particular model. And it already should be done. Here:

https://github.com/openvinotoolkit/openvino/blob/master/src/core/include/openvino/core/model.hpp#L46

File: src/cmake/ie_parallel.cmake (hunk 11 above)
Comment (ilya-lavrenov) [reply src/cmake/ie_parallel.cmake:120]: just put definition to `ie_parallel.hpp` as well :)

File: src/core/include/openvino/core/model.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply src/core/include/openvino/core/model.hpp:118]: we don't need to keep frontend manager here, ov::Model already keeps a reference to a particular frontend

[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/include/openvino/core/model.hpp:118]: > In general we don't need to have FR manager here. We need to keep only one frontend which created this particular model. And it already should be done. Here:
> 
> https://github.com/openvinotoolkit/openvino/blob/master/src/core/include/openvino/core/model.hpp#L46

It cannot work, we will encounter problem in some CI test, sometimes frontend manager will released before ov::Model, which cause segment fault.

[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/include/openvino/core/model.hpp:118]: > we don't need to keep frontend manager here, ov::Model already keeps a reference to a particular frontend

I add such reference at ov::Model is to fix segment fault in CI test failure.

File: src/core/src/model.cpp (hunk 15 above)
Comment (riverlijunjie) [reply src/core/src/model.cpp:40]: @ilya-lavrenov I sent some details info to you by email, please check it, thanks!

File: src/core/include/openvino/op/constant.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/core/include/openvino/op/constant.hpp:719]: @ilya-lavrenov @ilyachur I have done some test to remove femgr reference from ov::op::constants, see test PR:https://github.com/openvinotoolkit/openvino/pull/12166
Test result shows that we still need keep this reference in ov::op::constants, some test failed, such as below 2 cases:
1) test_IENetwork.py
2) test_NGraph.py
You can reference to PR12166 CI result.

Anyhow I reproduce the same failure in local, such as test_NGraph.py

> ov2022@ov2022-Z370-AORUS-Gaming-5:~/workspace/capi/openvino/src/bindings/python/tests_compatibility$ python3 -m pytest -s test_inference_engine/test_NGraph.py -v
> ========================================================================================= test session starts ==========================================================================================
> platform linux -- Python 3.8.10, pytest-4.6.9, py-1.8.1, pluggy-0.13.0 -- /usr/bin/python3
> cachedir: .pytest_cache
> rootdir: /home/ov2022/workspace/capi/openvino/src/bindings/python
> plugins: anyio-3.5.0
> collected 6 items                                                                                                                                                                                      
> 
> test_inference_engine/test_NGraph.py::test_create_IENetwork_from_nGraph PASSED
> test_inference_engine/test_NGraph.py::test_get_IENetwork_from_nGraph PASSED
> test_inference_engine/test_NGraph.py::test_get_ops_from_IENetwork **Segmentation fault (core dumped)**

 

The failure test item is:

> def test_get_ops_from_IENetwork():
>     ie = IECore()
>     net = ie.read_network(model=test_net_xml, weights=test_net_bin)
>     func = ng.function_from_cnn(net)
>     ops = func.get_ordered_ops()
>     ops_names = [op.friendly_name for op in ops]
>     assert len(ops_names) != 0
>     assert 'data' in ops_names

It's obvious that ov::Model has been released before ops, I dump the call trace as below:

test_inference_engine/test_NGraph.py ..

> Thread 1 "python3" received signal SIGSEGV, Segmentation fault.
> 0x00007fff939b03ba in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x1266dc0) at /usr/include/c++/9/bits/shared_ptr_base.h:155
> 155                 _M_dispose();
> (gdb) bt
> #0  0x00007fff939b03ba in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x1266dc0) at /usr/include/c++/9/bits/shared_ptr_base.h:155
> #1  0x00007fff939a9b95 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7fffffff7908, __in_chrg=<optimized out>) at /usr/include/c++/9/bits/shared_ptr_base.h:730
> #2  0x00007fff92924f10 in std::__shared_ptr<ngraph::runtime::AlignedBuffer, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fffffff7900, __in_chrg=<optimized out>)
>     at /usr/include/c++/9/bits/shared_ptr_base.h:1169
> #3  0x00007fff92940612 in std::__shared_ptr<ngraph::runtime::AlignedBuffer, (__gnu_cxx::_Lock_policy)2>::operator= (this=0x1267760, __r=...) at /usr/include/c++/9/bits/shared_ptr_base.h:1265
> #4  0x00007fff929382fa in std::shared_ptr<ngraph::runtime::AlignedBuffer>::operator= (this=0x1267760, __r=...) at /usr/include/c++/9/bits/shared_ptr.h:335
> #5  0x00007fff9292f5e0 in ov::op::v0::Constant::~Constant (this=0x1267570, __in_chrg=<optimized out>) at /home/ov2022/workspace/capi/openvino/src/core/src/op/constant.cpp:243
> #6  0x00007fff923ba647 in __gnu_cxx::new_allocator<ov::op::v0::Constant>::destroy<ov::op::v0::Constant> (this=0x1267570, __p=0x1267570) at /usr/include/c++/9/ext/new_allocator.h:152
> #7  0x00007fff923b9c31 in std::allocator_traits<std::allocator<ov::op::v0::Constant> >::destroy<ov::op::v0::Constant> (__a=..., __p=0x1267570) at /usr/include/c++/9/bits/alloc_traits.h:496
> #8  0x00007fff923b916d in std::_Sp_counted_ptr_inplace<ov::op::v0::Constant, std::allocator<ov::op::v0::Constant>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (this=0x1267560)
>     at /usr/include/c++/9/bits/shared_ptr_base.h:557
> #9  0x00007fff939b03c6 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x1267560) at /usr/include/c++/9/bits/shared_ptr_base.h:155
> #10 0x00007fff939a9b95 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7fff90685880, __in_chrg=<optimized out>) at /usr/include/c++/9/bits/shared_ptr_base.h:730
> #11 0x00007fff93a5563a in std::__shared_ptr<ov::op::v0::Constant, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fff90685878, __in_chrg=<optimized out>)
>     at /usr/include/c++/9/bits/shared_ptr_base.h:1169
> #12 0x00007fff93a5565a in std::shared_ptr<ov::op::v0::Constant>::~shared_ptr (this=0x7fff90685878, __in_chrg=<optimized out>) at /usr/include/c++/9/bits/shared_ptr.h:103
> #13 0x00007fff93a5ca2c in pybind11::class_<ov::op::v0::Constant, std::shared_ptr<ov::op::v0::Constant>, ov::Node>::dealloc (v_h=...)
>     at /home/ov2022/workspace/capi/openvino/src/bindings/python/thirdparty/pybind11/include/pybind11/pybind11.h:1612
> #14 0x00007fff9398f7a1 in pybind11::detail::clear_instance (self=<_pyngraph.op.Constant at remote 0x7fff90685860>)
>     at /home/ov2022/workspace/capi/openvino/src/bindings/python/thirdparty/pybind11/include/pybind11/detail/class.h:399
> #15 0x00007fff9398f8a3 in pybind11::detail::pybind11_object_dealloc (self=<_pyngraph.op.Constant at remote 0x7fff90685860>)
>     at /home/ov2022/workspace/capi/openvino/src/bindings/python/thirdparty/pybind11/include/pybind11/detail/class.h:419
> #16 0x00000000005ee3fb in _Py_Dealloc (op=<optimized out>) at ../Objects/object.c:2215
> #17 _Py_DECREF (filename=<synthetic pointer>, lineno=541, op=<optimized out>) at ../Include/object.h:478
> #18 _Py_XDECREF (op=<optimized out>) at ../Include/object.h:541
> #19 list_dealloc (op=0x7fff9060eb40) at ../Objects/listobject.c:372
> #20 0x00000000005ee8c0 in _Py_Dealloc (op=<optimized out>) at ../Objects/object.c:2215
> #21 _Py_DECREF (filename=0x88c960 "../Objects/frameobject.c", lineno=430, op=<optimized out>) at ../Include/object.h:478

File: src/core/include/openvino/core/model.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (ilyachur) [reply src/core/include/openvino/core/model.hpp:118]: > > In general we don't need to have FR manager here. We need to keep only one frontend which created this particular model. And it already should be done. Here:
> > https://github.com/openvinotoolkit/openvino/blob/master/src/core/include/openvino/core/model.hpp#L46
> 
> It cannot work, we will encounter problem in some CI test, sometimes frontend manager will released before ov::Model, which cause segment fault.

Can we investigate these issues in order to understand how frontend manager affects `ov::Model`? Frontend manager doesn't create `ov::Model`, it creates frontends and frontend creates `ov::Model`, it means that `ov::Model` depends on frontend but doesn't depend on FE manager.

File: src/core/include/openvino/op/constant.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (ilyachur) [reply src/core/include/openvino/op/constant.hpp:719]: I think it is not an issue of constant op. You can have the same problem if frontend creates `ov::Tensor`. we need to fix the issue for common case, but it can require deeper refactoring of existed structures.
